-1

**Failed Tests**:   Build  ClangBuild 
**Size**: This PR adds an extra 52KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c1121f/54702/summary.html
**COMMIT**: 77b81cd07de2105ff57a124dcb6bf2bb88095f5f
**CMSSW**: CMSSW_20_1_X_2026-07-12-0000/el9_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/51470/54702/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @bfonta cms-sw/cmssw#51006
 - @felicepantaleo cms-sw/cmssw#51213
 - @thomreis cms-sw/cmssw#51269
 - @smuzaffar cms-sw/cmssw#51375
 - @bsunanda cms-sw/cmssw#51463

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c1121f/54702/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c1121f/54702/git-merge-result


## Failed Build

I found compilation error when building: 

<pre>------- copying files from src/PhysicsTools/TruthInfo/scripts -------
>> copied makeTruthGraphValidationPlots.py
>> Compiling  src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/gcc/13.4.0-6908cfdf803923e783448096ca4f0923/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=130400 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DCMSSW_GIT_HASH='CMSSW_20_1_X_2026-07-12-0000' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_20_1_X_2026-07-12-0000' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/cms/cmssw/CMSSW_20_1_X_2026-07-12-0000/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/hepmc3/3.3.1-0e3a14abf956ab6455e0a82d19f54f77/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/pcre/8.43-7f0edb13c32a66555b25b5433dd2ceed/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/boost/1.91.0-4eabf1c007e1fd684a3ea24705495812/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/bz2lib/1.0.8-353e2ef1c2f1d06b28a8938defa197bc/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/clhep/2.4.7.2-4a01a851ada8a481b4c8f9856a60bb0b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/cppunit/1.15.x-08da40e4de8fd94fcb7bac1dce1b87fc/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/gsl/2.6-f7a733435831708073d54bcef0bb3ab2/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/hepmc/2.06.10-00a0d35b3057f230b67188fb41fc8f5e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/heppdt/3.04.01-aa4f6cddf7c53decf34691f13b2307b1/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/hls/2025.05-1e9bc123eaf0d9d12de4203b234629e6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/libuuid/2.40-d94705b84648fd294d63bf9de23bc03b/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/lcg/root/6.36.13-1e2cf577544a16a0d90570e219cd8156/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/tbb/v2022.3.0-dc50d53ffc8b6a202ff6e276cd3c9d79/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/xz/5.6.4-ea99661ac7ca423b4d74fd650284b41f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/zlib/1.3.2-53c061acc706fe8f28fd9477d9981c20/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/zstd/1.5.7-3d33d105dce0061a2affabfbb156d655/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/alpaka/2.1.1-3caaac8d71f39d400ab2511b2403675a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/eigen/c1d637433e3b3f9012b226c2c9125c494b470ae6-64a84c2742d1bb0d777807c58fc50caf/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/eigen/c1d637433e3b3f9012b226c2c9125c494b470ae6-64a84c2742d1bb0d777807c58fc50caf/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/fmt/10.2.1-2fdb9573b12b6a8373d9a65eba3c8c59/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/md5/1.0.0-26057075013e190e56dad37d35219376/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/OpenBLAS/0.3.27-47a3c8350874a1ad9f41674752b30581/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02950/el9_amd64_gcc13/external/tinyxml2/6.2.0-da2efb8ba86adb2f29c708e411784949/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=unused-value -Werror=unused-result -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DALPAKA_HAS_STD_ATOMIC_REF -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el9_amd64_gcc13/src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t/BranchHitAssociator_t.cpp.d src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp -o tmp/el9_amd64_gcc13/src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t/BranchHitAssociator_t.cpp.o
src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp: In function 'truth::LogicalGraphHitIndex {anonymous}::buildIndex()':
src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp:29:29: error: no matching function for call to 'truth::LogicalGraphHitIndexBuilder::setSimTrackForParticle(int, int)'
   29 |     b.setSimTrackForParticle(0, 100);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
In file included from src/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp:12:
src/PhysicsTools/TruthInfo/interface/LogicalGraphHitIndexBuilder.h:23:10: note: candidate: 'void truth::LogicalGraphHitIndexBuilder::setSimTrackForParticle(uint32_t, uint64_t, uint32_t)'
   23 |     void setSimTrackForParticle(uint32_t particleId, uint64_t eventId, uint32_t trackId);
</pre>

## Failed Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package SimG4Core/Application
>> Entering Package SimG4Core/Notification
>> Entering Package PhysicsTools/TruthInfo
>> Entering Package SimDataFormats/TruthInfo
>> Compile sequence completed for CMSSW CMSSW_20_1_X_2026-07-12-0000
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
Command exited with non-zero status 1
	Command being timed: "scram build -k -j 32 COMPILER=llvm compile BUILD_LOG=yes"
	User time (seconds): 349.25
	System time (seconds): 43.22
	Percent of CPU this job got: 1291%
</pre>

--- a/PhysicsTools/TruthInfo/interface/LogicalGraphHitIndexBuilder.h
+++ b/PhysicsTools/TruthInfo/interface/LogicalGraphHitIndexBuilder.h
@@ -18,17 +18,26 @@ namespace truth {
   public:
     explicit LogicalGraphHitIndexBuilder(uint32_t nParticles);
 
-    void setSimTrackForParticle(uint32_t particleId, uint32_t trackId);
+    // trackId is event-local (each mixing sub-event reuses 1,2,3,...); it MUST be
+    // namespaced by the packed EncodedEventId or signal and pileup collide.
+    void setSimTrackForParticle(uint32_t particleId, uint64_t eventId, uint32_t trackId);
     void addParticleChild(uint32_t parentParticleId, uint32_t childParticleId);
 
     // Add a hit on `trackId`'s SimTrack to `channel`. recHitIndex defaults to "no
     // recHit" for channels without a DetId->RecHit link (tracker, muon); calo/MTD
     // pass the mapped global recHit index.
     void addHit(HitChannel channel,
+                uint64_t eventId,
                 uint32_t trackId,
                 uint32_t detId,
                 float energy,
                 uint32_t recHitIndex = LogicalGraphHitIndex::Hit::kInvalidRecHitIndex);
+
+    // (EncodedEventId, trackId) -> global map key. The packed EncodedEventId fits in
+    // 32 bits (reco::EncodedEventId::rawId is uint32), so shift it into the high word.
+    static uint64_t simKey(uint64_t eventId, uint32_t trackId) {
+      return (eventId << 32) | static_cast<uint64_t>(trackId);
+    }
 
     [[nodiscard]] LogicalGraphHitIndex finish();
 
@@ -72,7 +81,7 @@ namespace truth {
 
     uint32_t nParticles_ = 0;
 
-    std::unordered_map<uint32_t, uint32_t> trackIdToParticle_;
+    std::unordered_map<uint64_t, uint32_t> trackIdToParticle_;
     std::vector<std::vector<uint32_t>> children_;
 
     // [channel index][particle] -> direct hit list. Subgraph hits are aggregated

--- a/PhysicsTools/TruthInfo/plugins/BuildFile.xml
+++ b/PhysicsTools/TruthInfo/plugins/BuildFile.xml
@@ -34,7 +34,6 @@
 <use name="DataFormats/ParticleFlowReco"/>
 <use name="SimDataFormats/TrackingHit"/>
 <use name="Geometry/TrackerGeometryBuilder"/>
-<use name="Geometry/CommonDetUnit"/>
 <use name="Geometry/CommonTopologies"/>
 
 <use name="DataFormats/TrackReco"/>

--- a/PhysicsTools/TruthInfo/plugins/LogicalGraphHitIndexProducer.cc
+++ b/PhysicsTools/TruthInfo/plugins/LogicalGraphHitIndexProducer.cc
@@ -336,7 +336,7 @@ void TruthLogicalGraphHitIndexProducer::fillTrackToParticleMap(LogicalGraphView 
     if (trackId == 0)
       continue;
 
-    builder.setSimTrackForParticle(particleId, trackId);
+    builder.setSimTrackForParticle(particleId, rawGraph.nodeEventId(simNodeU32), trackId);
   }
 
   for (uint32_t parentId = 0; parentId < graph.nParticles(); ++parentId) {
@@ -487,8 +487,12 @@ void TruthLogicalGraphHitIndexProducer::fillSimHits(edm::Event& event,
         }
       }
 
-      builder.addHit(
-          truth::HitChannel::HGCalCalo, static_cast<uint32_t>(geantTrackId), detId, simHit.energy(), recHitIndex);
+      builder.addHit(truth::HitChannel::HGCalCalo,
+                     simHit.eventId().rawId(),
+                     static_cast<uint32_t>(geantTrackId),
+                     detId,
+                     simHit.energy(),
+                     recHitIndex);
     }
   }
 }
@@ -508,7 +512,11 @@ void TruthLogicalGraphHitIndexProducer::fillTrackerSimHits(edm::Event& event,
     for (auto const& simHit : *hSimHits) {
       // PSimHit::trackId() is the G4 trackId of the SimTrack that made the hit,
       // the same id space used to associate calorimeter simhits to particles.
-      builder.addHit(truth::HitChannel::Tracker, simHit.trackId(), simHit.detUnitId(), simHit.energyLoss());
+      builder.addHit(truth::HitChannel::Tracker,
+                     simHit.eventId().rawId(),
+                     simHit.trackId(),
+                     simHit.detUnitId(),
+                     simHit.energyLoss());
     }
   }
 }
@@ -524,7 +532,8 @@ void TruthLogicalGraphHitIndexProducer::fillMuonSimHits(edm::Event& event,
       continue;
 
     for (auto const& simHit : *hSimHits) {
-      builder.addHit(truth::HitChannel::Muon, simHit.trackId(), simHit.detUnitId(), simHit.energyLoss());
+      builder.addHit(
+          truth::HitChannel::Muon, simHit.eventId().rawId(), simHit.trackId(), simHit.detUnitId(), simHit.energyLoss());
     }
   }
 }
@@ -578,7 +587,11 @@ void TruthLogicalGraphHitIndexProducer::fillMtdHits(edm::Event& event,
     const auto trackId = static_cast<uint32_t>(cluster.particleId());
     for (auto const& [packedHit, energy] : cluster.hits_and_energies()) {
       const uint32_t moduleDetId = static_cast<uint32_t>(packedHit >> 32);
-      builder.addHit(truth::HitChannel::MTD, trackId, moduleDetId, energy, recHitIndex);
+      // eventId 0 is correct only because the clusters above are filtered to the signal
+      // interaction (bx 0, event 0), so simKey(0, trackId) matches the signal track.
+      // If a merged pileup MTD collection is ever added, this must pass the real
+      // eventId like the calo/tracker channels or pileup tracks would alias the signal.
+      builder.addHit(truth::HitChannel::MTD, 0ull, trackId, moduleDetId, energy, recHitIndex);
     }
   }
 }

--- a/PhysicsTools/TruthInfo/plugins/TruthGraphAccumulator.cc
+++ b/PhysicsTools/TruthInfo/plugins/TruthGraphAccumulator.cc
@@ -40,9 +40,12 @@
 #include "FWCore/Framework/interface/ProducesCollector.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/Utilities/interface/InputTag.h"
 #include "FWCore/Utilities/interface/StreamID.h"
 
+#include "SimDataFormats/CaloHit/interface/PCaloHit.h"
+#include "SimDataFormats/TrackingHit/interface/PSimHit.h"
 #include "SimGeneral/MixingModule/interface/DigiAccumulatorMixMod.h"
 #include "SimGeneral/MixingModule/interface/DigiAccumulatorMixModFactory.h"
 #include "SimGeneral/MixingModule/interface/PileUpEventPrincipal.h"
@@ -62,9 +65,10 @@
 
 namespace {
   uint64_t packEventId(EncodedEventId const& id) {
-    uint64_t out = 0;
-    std::memcpy(&out, &id, sizeof(EncodedEventId));
-    return out;
+    // EncodedEventId is a single uint32 rawId; use the typed accessor rather than a
+    // byte copy so the key stays portable and cannot pick up a future member/padding.
+    static_assert(sizeof(EncodedEventId) == sizeof(uint32_t));
+    return static_cast<uint64_t>(id.rawId());
   }
 
   // Stable (status 1) GEN particles as (barcode, pdgId). Used to collapse the GEN
@@ -124,15 +128,49 @@ private:
                    edm::SimVertexContainer const& vertices,
                    EncodedEventId const& eid);
 
+  // Append this sub-event's sim-hits to the merged collections, re-tagged with `eid`
+  // so they carry per-interaction provenance (native hits are all tagged (0,0)).
+  template <class EvT>
+  void addSubEventHits(EvT const& ev, EncodedEventId const& eid);
+
+  // Merge one sim-hit collection family (PCaloHit or PSimHit) from the sub-event,
+  // re-tagging each hit's eventId. Kept per subdetector family so a downstream
+  // consumer can apply the right sim-to-reco DetId relabelling per collection.
+  template <class HitT, class EvT>
+  void mergeHits(EvT const& ev,
+                 std::vector<edm::InputTag> const& tags,
+                 EncodedEventId const& eid,
+                 std::vector<HitT>& out);
+
   const edm::InputTag simTrackTag_;
   const edm::InputTag simVertexTag_;
   const edm::InputTag hepmc3Tag_;
   const edm::InputTag hepmc2Tag_;
+  const std::vector<edm::InputTag> caloHitTags_;
+  const std::vector<edm::InputTag> ecalHitTags_;
+  const std::vector<edm::InputTag> hcalHitTags_;
+  const std::vector<edm::InputTag> trackerHitTags_;
+  const std::vector<edm::InputTag> muonHitTags_;
+  const std::vector<edm::InputTag> mtdHitTags_;
   const std::vector<int> pileupBunchCrossings_;
   const bool collapsePileupGen_;
   const bool collapseSignalGen_;
 
-  std::unordered_map<int, int> pileupCountByBx_;
+  int pileupCount_ = 0;
+  bool missingCaloHitsWarned_ = false;
+
+  // Merged calorimeter sim-hits across signal + kept pileup, each re-tagged with its
+  // sub-event EncodedEventId so the (eventId,trackId) hit-index key resolves pileup
+  // nodes at RECO (the native pileup hits are consumed transiently here). Kept one
+  // vector per subdetector family so the relabelling at RECO stays per collection.
+  std::vector<PCaloHit> mergedCaloHits_;
+  std::vector<PCaloHit> mergedEcalHits_;
+  std::vector<PCaloHit> mergedHcalHits_;
+  // Tracking sim-hits (tracker, muon chambers, MTD) as PSimHit, same per-interaction
+  // re-tagging. Tracker pileup is by far the largest family; see the customise note.
+  std::vector<PSimHit> mergedTrackerHits_;
+  std::vector<PSimHit> mergedMuonHits_;
+  std::vector<PSimHit> mergedMtdHits_;
 
   std::vector<TruthGraph::NodeRef> nodes_;
   std::vector<int32_t> pdgId_;
@@ -157,18 +195,42 @@ TruthGraphAccumulator::TruthGraphAccumulator(edm::ParameterSet const& cfg,
       simVertexTag_(cfg.getParameter<edm::InputTag>("simVertices")),
       hepmc3Tag_(cfg.getParameter<edm::InputTag>("genEventHepMC3")),
       hepmc2Tag_(cfg.getParameter<edm::InputTag>("genEventHepMC")),
+      caloHitTags_(cfg.getParameter<std::vector<edm::InputTag>>("caloHits")),
+      ecalHitTags_(cfg.getParameter<std::vector<edm::InputTag>>("ecalHits")),
+      hcalHitTags_(cfg.getParameter<std::vector<edm::InputTag>>("hcalHits")),
+      trackerHitTags_(cfg.getParameter<std::vector<edm::InputTag>>("trackerHits")),
+      muonHitTags_(cfg.getParameter<std::vector<edm::InputTag>>("muonHits")),
+      mtdHitTags_(cfg.getParameter<std::vector<edm::InputTag>>("mtdHits")),
       pileupBunchCrossings_(cfg.getParameter<std::vector<int>>("pileupBunchCrossings")),
       collapsePileupGen_(cfg.getParameter<bool>("collapsePileupGen")),
       collapseSignalGen_(cfg.getParameter<bool>("collapseSignalGen")) {
   producesCollector.produces<TruthGraph>();
+  producesCollector.produces<std::vector<PCaloHit>>("mergedHGCHits");
+  producesCollector.produces<std::vector<PCaloHit>>("mergedEcalHits");
+  producesCollector.produces<std::vector<PCaloHit>>("mergedHcalHits");
+  producesCollector.produces<std::vector<PSimHit>>("mergedTrackerHits");
+  producesCollector.produces<std::vector<PSimHit>>("mergedMuonHits");
+  producesCollector.produces<std::vector<PSimHit>>("mergedMtdHits");
   iC.consumes<edm::SimTrackContainer>(simTrackTag_);
   iC.consumes<edm::SimVertexContainer>(simVertexTag_);
   iC.mayConsume<edm::HepMC3Product>(hepmc3Tag_);
   iC.mayConsume<edm::HepMCProduct>(hepmc2Tag_);
+  for (auto const* tags : {&caloHitTags_, &ecalHitTags_, &hcalHitTags_})
+    for (auto const& tag : *tags)
+      iC.mayConsume<std::vector<PCaloHit>>(tag);
+  for (auto const* tags : {&trackerHitTags_, &muonHitTags_, &mtdHitTags_})
+    for (auto const& tag : *tags)
+      iC.mayConsume<std::vector<PSimHit>>(tag);
 }
 
 void TruthGraphAccumulator::initializeEvent(edm::Event const&, edm::EventSetup const&) {
-  pileupCountByBx_.clear();
+  pileupCount_ = 0;
+  mergedCaloHits_.clear();
+  mergedEcalHits_.clear();
+  mergedHcalHits_.clear();
+  mergedTrackerHits_.clear();
+  mergedMuonHits_.clear();
+  mergedMtdHits_.clear();
   nodes_.clear();
   pdgId_.clear();
   status_.clear();
@@ -270,6 +332,44 @@ void TruthGraphAccumulator::addSubEvent(std::vector<std::pair<int, int>> const& 
   }
 }
 
+template <class HitT, class EvT>
+void TruthGraphAccumulator::mergeHits(EvT const& ev,
+                                      std::vector<edm::InputTag> const& tags,
+                                      EncodedEventId const& eid,
+                                      std::vector<HitT>& out) {
+  for (auto const& tag : tags) {
+    edm::Handle<std::vector<HitT>> hits;
+    ev.getByLabel(tag, hits);
+    if (!hits.isValid()) {
+      // Under premixed pileup the pileup sim-hits are already digitized away, so every
+      // pileup handle is invalid and the merged collection ends up signal-only, silently
+      // reverting the pileup-aware truth to signal-only. Warn once.
+      if (!missingCaloHitsWarned_) {
+        edm::LogWarning("TruthGraphAccumulator")
+            << "sim-hit collection " << tag.encode()
+            << " not found for a sub-event; pileup-aware truth needs classic (non-premixed) pileup.";
+        missingCaloHitsWarned_ = true;
+      }
+      continue;
+    }
+    out.reserve(out.size() + hits->size());
+    for (HitT hit : *hits) {  // copy: re-tag the eventId to this sub-event
+      hit.setEventId(eid);
+      out.push_back(hit);
+    }
+  }
+}
+
+template <class EvT>
+void TruthGraphAccumulator::addSubEventHits(EvT const& ev, EncodedEventId const& eid) {
+  mergeHits(ev, caloHitTags_, eid, mergedCaloHits_);
+  mergeHits(ev, ecalHitTags_, eid, mergedEcalHits_);
+  mergeHits(ev, hcalHitTags_, eid, mergedHcalHits_);
+  mergeHits(ev, trackerHitTags_, eid, mergedTrackerHits_);
+  mergeHits(ev, muonHitTags_, eid, mergedMuonHits_);
+  mergeHits(ev, mtdHitTags_, eid, mergedMtdHits_);
+}
+
 void TruthGraphAccumulator::accumulate(edm::Event const& event, edm::EventSetup const&) {
   edm::Handle<edm::SimTrackContainer> tracks;
   edm::Handle<edm::SimVertexContainer> vertices;
@@ -280,7 +380,9 @@ void TruthGraphAccumulator::accumulate(edm::Event const& event, edm::EventSetup 
   std::vector<std::pair<int, int>> stableGen;
   if (collapseSignalGen_)
     stableGen = readStableGen(event, hepmc3Tag_, hepmc2Tag_);
-  addSubEvent(stableGen, *tracks, *vertices, EncodedEventId(0, 0));
+  const EncodedEventId sigEid(0, 0);
+  addSubEvent(stableGen, *tracks, *vertices, sigEid);
+  addSubEventHits(event, sigEid);
 }
 
 void TruthGraphAccumulator::accumulate(PileUpEventPrincipal const& pep, edm::EventSetup const&, edm::StreamID const&) {
@@ -299,8 +401,18 @@ void TruthGraphAccumulator::accumulate(PileUpEventPrincipal const& pep, edm::Eve
   if (collapsePileupGen_)
     stableGen = readStableGen(pep, hepmc3Tag_, hepmc2Tag_);
 
-  const int puIndex = ++pileupCountByBx_[bx];
-  addSubEvent(stableGen, *tracks, *vertices, EncodedEventId(bx, puIndex));
+  // Global counter across bunch crossings: EncodedEventId stores abs(bx), so a
+  // per-bx counter would give (-1,1) and (+1,1) identical packed ids. A single
+  // counter keeps every pileup interaction's tag unique regardless of bx sign.
+  const int puIndex = ++pileupCount_;
+  // EncodedEventId packs the event number into 16 bits; an unrealistic pileup
+  // multiplicity would overflow into the bunch-crossing bits and alias ids.
+  if (puIndex > 0xFFFF)
+    throw cms::Exception("TruthGraphAccumulator")
+        << "pileup sub-event count " << puIndex << " exceeds the 16-bit EncodedEventId event field";
+  const EncodedEventId puEid(bx, puIndex);
+  addSubEvent(stableGen, *tracks, *vertices, puEid);
+  addSubEventHits(pep, puEid);
 }
 
 void TruthGraphAccumulator::finalizeEvent(edm::Event& event, edm::EventSetup const&) {
@@ -340,6 +452,13 @@ void TruthGraphAccumulator::finalizeEvent(edm::Event& event, edm::EventSetup con
     throw cms::Exception("TruthGraphAccumulator") << "Produced TruthGraph is not consistent";
 
   event.put(std::move(out));
+
+  event.put(std::make_unique<std::vector<PCaloHit>>(std::move(mergedCaloHits_)), "mergedHGCHits");
+  event.put(std::make_unique<std::vector<PCaloHit>>(std::move(mergedEcalHits_)), "mergedEcalHits");
+  event.put(std::make_unique<std::vector<PCaloHit>>(std::move(mergedHcalHits_)), "mergedHcalHits");
+  event.put(std::make_unique<std::vector<PSimHit>>(std::move(mergedTrackerHits_)), "mergedTrackerHits");
+  event.put(std::make_unique<std::vector<PSimHit>>(std::move(mergedMuonHits_)), "mergedMuonHits");
+  event.put(std::make_unique<std::vector<PSimHit>>(std::move(mergedMtdHits_)), "mergedMtdHits");
 }
 
 DEFINE_DIGI_ACCUMULATOR(TruthGraphAccumulator);

--- a/PhysicsTools/TruthInfo/plugins/TruthLogicalGraphProducer.cc
+++ b/PhysicsTools/TruthInfo/plugins/TruthLogicalGraphProducer.cc
@@ -888,7 +888,12 @@ public:
     std::vector<uint8_t> particleDirectHit;
 
     if (dropHitlessSimSubgraphs_) {
-      std::unordered_set<uint32_t> hitTrackIds;
+      // trackId is event-local (each mixing sub-event reuses 1,2,3,...), so it MUST
+      // be namespaced by the packed EncodedEventId or signal and pileup collide and
+      // the wrong particles get flagged as hit-bearing. Mirrors the same key in
+      // LogicalGraphHitIndexBuilder so the pruned graph stays consistent with the index.
+      auto simKey = [](uint64_t eventId, uint32_t trackId) { return (eventId << 32) | static_cast<uint64_t>(trackId); };
+      std::unordered_set<uint64_t> hitKeys;
       bool anyCollectionValid = false;
 
       for (auto const& token : caloSimHitTokens_) {
@@ -900,7 +905,7 @@ public:
         for (auto const& hit : *hHits) {
           const int trackId = hit.geantTrackId();
           if (trackId > 0 && hit.energy() > 0.f)
-            hitTrackIds.insert(static_cast<uint32_t>(trackId));
+            hitKeys.insert(simKey(hit.eventId().rawId(), static_cast<uint32_t>(trackId)));
         }
       }
 
@@ -912,7 +917,7 @@ public:
         anyCollectionValid = true;
         for (auto const& hit : *hHits) {
           if (hit.energyLoss() > 0.f)
-            hitTrackIds.insert(hit.trackId());
+            hitKeys.insert(simKey(hit.eventId().rawId(), hit.trackId()));
         }
       }
 
@@ -930,7 +935,7 @@ public:
             continue;
           if (ref.key <= 0 || ref.key > static_cast<int64_t>(std::numeric_limits<uint32_t>::max()))
             continue;
-          if (hitTrackIds.count(static_cast<uint32_t>(ref.key)) != 0)
+          if (hitKeys.count(simKey(raw.nodeEventId(simNodeU32), static_cast<uint32_t>(ref.key))) != 0)
             particleDirectHit[particleId] = 1;
         }
       } else {

--- a/PhysicsTools/TruthInfo/python/customiseDropIncompleteValidators.py
+++ b/PhysicsTools/TruthInfo/python/customiseDropIncompleteValidators.py
@@ -1,0 +1,25 @@
+"""Harness-only customise for producing HGCAL validation plots from standalone
+RECO-from-step2 jobs. It removes validators that need EventSetup records or
+collections a standalone offline job does not provide (HCAL trigger-primitive
+record, HLT-only reco), keeping the offline HGCal validator and its associators
+intact. Not for PR configs: the CI workflows run the full chain and provide
+these inputs."""
+
+
+def customise(process):
+    drop = [
+        "AllHcalDigisValidation",   # needs CaloTPGRecord
+        "hcalDigisValidationSequence",
+        "hltHgcalValidator",        # HLT TICL reco not run in an offline RECO job
+        "tpHltGsfTrackAssociation",
+        "hltGsfTrackValidator",
+        "hltTrackValidator",
+        "hltMultiTrackValidator",
+    ]
+    for label in drop:
+        if hasattr(process, label):
+            process.__delattr__(label)
+    # Also blank any path/sequence referencing the dropped labels leniently by
+    # rebuilding the schedule without empty leftovers is unnecessary: EDM prunes
+    # unscheduled modules, and removing the module makes the sequence skip it.
+    return process

--- a/PhysicsTools/TruthInfo/python/customiseTruthMixedReco.py
+++ b/PhysicsTools/TruthInfo/python/customiseTruthMixedReco.py
@@ -1,0 +1,50 @@
+"""RECO-side customise for the pileup truth chain: build the merged (signal+pileup)
+logical graph from the MixingModule accumulator's raw TruthGraph (label mix) and
+resolve its SimHit associations against the mixed rechits. Pairs with
+mixedTruthGraphCustomize.addTruthGraphAccumulator at DIGI, giving a pileup-aware
+truth graph at RECO."""
+
+import FWCore.ParameterSet.Config as cms
+
+
+def customise(process):
+    from PhysicsTools.TruthInfo.truthGraphValidation_cff import (
+        truthLogicalGraphProducer,
+        detIdToRecHitMapProducer,
+        truthLogicalGraphHitIndexProducer,
+    )
+    # The merged raw TruthGraph comes from the mixing accumulator, not the
+    # standalone truthGraphProducer.
+    process.truthLogicalGraphProducer = truthLogicalGraphProducer.clone(
+        src=cms.InputTag("mix"),
+        # The hitless-subgraph pruning decides which SimTracks left a calo hit; feed it
+        # the merged (signal+pileup) HGCal sim-hits, else every pileup subgraph looks
+        # hitless (its hits are in mix:mergedHGCHits, not the signal-only g4SimHits) and
+        # is pruned, leaving pileup branches empty. Matches the hit index below.
+        simHitCollections=cms.VInputTag(cms.InputTag('mix', 'mergedHGCHits')),
+    )
+    process.detIdToRecHitMapProducer = detIdToRecHitMapProducer
+    # The hit index also reads the RAW merged graph (for trackId->node); point it at mix.
+    process.truthLogicalGraphHitIndexProducer = truthLogicalGraphHitIndexProducer.clone(
+        rawSrc=cms.InputTag('mix'),
+        # Read the merged (signal+pileup) HGCal sim-hits from the accumulator, each
+        # tagged with its sub-event EncodedEventId, instead of the signal-only
+        # g4SimHits (which lack pileup at RECO).
+        simHitCollections=cms.VInputTag(cms.InputTag('mix', 'mergedHGCHits')),
+    )
+
+    process.truthMixedRecoPath = cms.Path(
+        process.truthLogicalGraphProducer
+        + process.detIdToRecHitMapProducer
+        + process.truthLogicalGraphHitIndexProducer
+    )
+    process.schedule.append(process.truthMixedRecoPath)
+
+    for out in process.outputModules_().values():
+        out.outputCommands.extend([
+            "keep *_truthLogicalGraphProducer_*_*",
+        ])
+        # The hit index is an intermediate graph footprint, regenerable from the graph
+        # plus mix:mergedHGCHits; persisting it costs ~9.7 MB/event (subgraph-hit CSR
+        # over every retained pileup particle), so it is kept out of the event content.
+    return process

--- a/PhysicsTools/TruthInfo/python/mixedTruthGraphCustomize.py
+++ b/PhysicsTools/TruthInfo/python/mixedTruthGraphCustomize.py
@@ -33,20 +33,55 @@ def addMixedTruthGraph(process):
 
 def addTruthGraphAccumulator(process,
                              pileupBunchCrossings=(0,),
-                             collapsePileupGen=True):
+                             collapsePileupGen=True,
+                             includeTrackingHits=False):
     """Phase-B (B1): register TruthGraphAccumulator inside the MixingModule.
 
     The accumulator builds the mixed (signal + pileup) raw TruthGraph from the
     native per-sub-event SimTrack/SimVertex collections. By default only in-time
     pileup (bx 0) is included; pass pileupBunchCrossings to widen. The mixed graph
     is kept in the output as TruthGraph_mix__<process>.
+
+    By default the accumulator captures only the CALORIMETER sim-hits (HGCAL plus
+    barrel ECAL/HCAL). The tracking detectors (tracker, muon chambers, MTD) are the
+    largest sim-hit family at PU200 and dominate the event size (the merged tracker
+    PSimHits alone are tens of MB/event), so they are OFF by default. Pass
+    includeTrackingHits=True for a full-detector truth graph.
     """
+    def tags(*names):
+        return cms.VInputTag(*[cms.InputTag("g4SimHits", n) for n in names])
+
+    # Tracking detectors only when explicitly requested; empty (calo-only) otherwise.
+    trackerHits = cms.VInputTag()
+    muonHits = cms.VInputTag()
+    mtdHits = cms.VInputTag()
+    if includeTrackingHits:
+        trackerHits = tags(
+            "TrackerHitsPixelBarrelLowTof", "TrackerHitsPixelBarrelHighTof",
+            "TrackerHitsPixelEndcapLowTof", "TrackerHitsPixelEndcapHighTof",
+            "TrackerHitsTIBLowTof", "TrackerHitsTIBHighTof",
+            "TrackerHitsTIDLowTof", "TrackerHitsTIDHighTof",
+            "TrackerHitsTOBLowTof", "TrackerHitsTOBHighTof",
+            "TrackerHitsTECLowTof", "TrackerHitsTECHighTof",
+        )
+        muonHits = tags("MuonDTHits", "MuonCSCHits", "MuonRPCHits", "MuonGEMHits", "MuonME0Hits")
+        mtdHits = tags("FastTimerHitsBarrel", "FastTimerHitsEndcap")
+
     process.mix.digitizers.truthGraph = cms.PSet(
         accumulatorType=cms.string("TruthGraphAccumulator"),
         simTracks=cms.InputTag("g4SimHits"),
         simVertices=cms.InputTag("g4SimHits"),
         genEventHepMC3=cms.InputTag("generatorSmeared"),
         genEventHepMC=cms.InputTag("generatorSmeared"),
+        caloHits=tags("HGCHitsEE", "HGCHitsHEfront", "HGCHitsHEback"),
+        # Barrel calorimeters, kept in separate products so the RECO consumer applies
+        # the right sim-to-reco DetId relabelling per collection (ECAL barrel needs
+        # none, HCAL uses HcalHitRelabeller).
+        ecalHits=tags("EcalHitsEB"),
+        hcalHits=tags("HcalHits"),
+        trackerHits=trackerHits,
+        muonHits=muonHits,
+        mtdHits=mtdHits,
         pileupBunchCrossings=cms.vint32(*pileupBunchCrossings),
         collapsePileupGen=cms.bool(collapsePileupGen),
         collapseSignalGen=cms.bool(False),
@@ -54,5 +89,16 @@ def addTruthGraphAccumulator(process,
 
     for out in process.outputModules_().values():
         out.outputCommands.append("keep TruthGraph_mix_*_*")
+        # The merged sim-hit collections are the union of signal + all kept pileup hits,
+        # and must bridge a split DIGI->RECO job (the pileup hits are gone after mixing;
+        # the RECO customise does not re-keep them). Drop these lines for a single-job
+        # DIGI+RECO where the hit index is built in the same process.
+        out.outputCommands.append("keep *_mix_mergedHGCHits_*")
+        out.outputCommands.append("keep *_mix_mergedEcalHits_*")
+        out.outputCommands.append("keep *_mix_mergedHcalHits_*")
+        if includeTrackingHits:
+            out.outputCommands.append("keep *_mix_mergedTrackerHits_*")
+            out.outputCommands.append("keep *_mix_mergedMuonHits_*")
+            out.outputCommands.append("keep *_mix_mergedMtdHits_*")
 
     return process

--- a/PhysicsTools/TruthInfo/src/LogicalGraphHitIndexBuilder.cc
+++ b/PhysicsTools/TruthInfo/src/LogicalGraphHitIndexBuilder.cc
@@ -16,11 +16,11 @@ namespace truth {
       channel.resize(nParticles);
   }
 
-  void LogicalGraphHitIndexBuilder::setSimTrackForParticle(uint32_t particleId, uint32_t trackId) {
+  void LogicalGraphHitIndexBuilder::setSimTrackForParticle(uint32_t particleId, uint64_t eventId, uint32_t trackId) {
     if (particleId >= nParticles_)
       return;
 
-    trackIdToParticle_[trackId] = particleId;
+    trackIdToParticle_[simKey(eventId, trackId)] = particleId;
   }
 
   void LogicalGraphHitIndexBuilder::addParticleChild(uint32_t parentParticleId, uint32_t childParticleId) {
@@ -31,11 +31,11 @@ namespace truth {
   }
 
   void LogicalGraphHitIndexBuilder::addHit(
-      HitChannel channel, uint32_t trackId, uint32_t detId, float energy, uint32_t recHitIndex) {
+      HitChannel channel, uint64_t eventId, uint32_t trackId, uint32_t detId, float energy, uint32_t recHitIndex) {
     if (energy <= 0.f)
       return;
 
-    auto it = trackIdToParticle_.find(trackId);
+    auto it = trackIdToParticle_.find(simKey(eventId, trackId));
     if (it == trackIdToParticle_.end())
       return;
 

--- a/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp
+++ b/PhysicsTools/TruthInfo/test/BranchHitAssociator_t.cpp
@@ -26,13 +26,13 @@ namespace {
   //    cellTotal   = {10:1, 11:2, 12:2}
   truth::LogicalGraphHitIndex buildIndex() {
     truth::LogicalGraphHitIndexBuilder b(2);
-    b.setSimTrackForParticle(0, 100);
-    b.setSimTrackForParticle(1, 101);
+    b.setSimTrackForParticle(0, 0, 100);
+    b.setSimTrackForParticle(1, 0, 101);
     b.addParticleChild(0, 1);
-    b.addHit(truth::HitChannel::HGCalCalo, 100, 10, 1.0f, 0);
-    b.addHit(truth::HitChannel::HGCalCalo, 100, 11, 1.0f, 0);
-    b.addHit(truth::HitChannel::HGCalCalo, 101, 11, 1.0f, 0);
-    b.addHit(truth::HitChannel::HGCalCalo, 101, 12, 2.0f, 0);
+    b.addHit(truth::HitChannel::HGCalCalo, 0, 100, 10, 1.0f, 0);
+    b.addHit(truth::HitChannel::HGCalCalo, 0, 100, 11, 1.0f, 0);
+    b.addHit(truth::HitChannel::HGCalCalo, 0, 101, 11, 1.0f, 0);
+    b.addHit(truth::HitChannel::HGCalCalo, 0, 101, 12, 2.0f, 0);
     return b.finish();
   }
 
@@ -40,14 +40,14 @@ namespace {
   // calo cell (10) that the tracker associator must ignore.
   truth::LogicalGraphHitIndex buildTrackerIndex() {
     truth::LogicalGraphHitIndexBuilder b(2);
-    b.setSimTrackForParticle(0, 100);
-    b.setSimTrackForParticle(1, 101);
+    b.setSimTrackForParticle(0, 0, 100);
+    b.setSimTrackForParticle(1, 0, 101);
     b.addParticleChild(0, 1);
-    b.addHit(truth::HitChannel::HGCalCalo, 100, 10, 1.0f, 0);  // calo channel
-    b.addHit(truth::HitChannel::Tracker, 100, 20, 1.0f);
-    b.addHit(truth::HitChannel::Tracker, 100, 21, 1.0f);
-    b.addHit(truth::HitChannel::Tracker, 101, 21, 1.0f);
-    b.addHit(truth::HitChannel::Tracker, 101, 22, 2.0f);
+    b.addHit(truth::HitChannel::HGCalCalo, 0, 100, 10, 1.0f, 0);  // calo channel
+    b.addHit(truth::HitChannel::Tracker, 0, 100, 20, 1.0f);
+    b.addHit(truth::HitChannel::Tracker, 0, 100, 21, 1.0f);
+    b.addHit(truth::HitChannel::Tracker, 0, 101, 21, 1.0f);
+    b.addHit(truth::HitChannel::Tracker, 0, 101, 22, 2.0f);
     return b.finish();
   }
 

--- a/PhysicsTools/TruthInfo/test/LogicalGraphHitIndexBuilder_t.cpp
+++ b/PhysicsTools/TruthInfo/test/LogicalGraphHitIndexBuilder_t.cpp
@@ -33,15 +33,15 @@ CPPUNIT_TEST_SUITE_REGISTRATION(TestLogicalGraphHitIndexBuilder);
 void TestLogicalGraphHitIndexBuilder::testSubgraphHitsAreSortedContiguousAndAccumulated() {
   // particle 0 (track 100) -> child particle 1 (track 101)
   truth::LogicalGraphHitIndexBuilder builder(2);
-  builder.setSimTrackForParticle(0, 100);
-  builder.setSimTrackForParticle(1, 101);
+  builder.setSimTrackForParticle(0, 0, 100);
+  builder.setSimTrackForParticle(1, 0, 101);
   builder.addParticleChild(0, 1);
 
-  builder.addHit(truth::HitChannel::HGCalCalo, 100, /*detId=*/10, /*energy=*/1.0f, /*recHitIndex=*/0);
-  builder.addHit(truth::HitChannel::HGCalCalo, 100, /*detId=*/5, /*energy=*/2.0f, /*recHitIndex=*/1);
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 100, /*detId=*/10, /*energy=*/1.0f, /*recHitIndex=*/0);
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 100, /*detId=*/5, /*energy=*/2.0f, /*recHitIndex=*/1);
   builder.addHit(
-      truth::HitChannel::HGCalCalo, 101, /*detId=*/10, /*energy=*/3.0f, /*recHitIndex=*/0);  // same detId as parent
-  builder.addHit(truth::HitChannel::HGCalCalo, 101, /*detId=*/20, /*energy=*/1.5f, /*recHitIndex=*/2);
+      truth::HitChannel::HGCalCalo, 0, 101, /*detId=*/10, /*energy=*/3.0f, /*recHitIndex=*/0);  // same detId as parent
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 101, /*detId=*/20, /*energy=*/1.5f, /*recHitIndex=*/2);
 
   auto index = builder.finish();
 
@@ -66,10 +66,10 @@ void TestLogicalGraphHitIndexBuilder::testSubgraphHitsAreSortedContiguousAndAccu
 
 void TestLogicalGraphHitIndexBuilder::testDirectHitsAreSortedByDetId() {
   truth::LogicalGraphHitIndexBuilder builder(1);
-  builder.setSimTrackForParticle(0, 7);
-  builder.addHit(truth::HitChannel::HGCalCalo, 7, 30, 1.0f, 0);
-  builder.addHit(truth::HitChannel::HGCalCalo, 7, 3, 1.0f, 1);
-  builder.addHit(truth::HitChannel::HGCalCalo, 7, 17, 1.0f, 2);
+  builder.setSimTrackForParticle(0, 0, 7);
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 7, 30, 1.0f, 0);
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 7, 3, 1.0f, 1);
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 7, 17, 1.0f, 2);
 
   auto index = builder.finish();
   auto direct = index.directHits(truth::HitChannel::HGCalCalo, 0);
@@ -85,16 +85,16 @@ void TestLogicalGraphHitIndexBuilder::testSubgraphDiamondCountsSharedDescendantO
   // once. (Regression: the old recursive child-subgraph merge summed it once per
   // path, since coalesce() sums equal detIds, doubling the energy.)
   truth::LogicalGraphHitIndexBuilder builder(4);
-  builder.setSimTrackForParticle(0, 100);
-  builder.setSimTrackForParticle(1, 101);
-  builder.setSimTrackForParticle(2, 102);
-  builder.setSimTrackForParticle(3, 103);
+  builder.setSimTrackForParticle(0, 0, 100);
+  builder.setSimTrackForParticle(1, 0, 101);
+  builder.setSimTrackForParticle(2, 0, 102);
+  builder.setSimTrackForParticle(3, 0, 103);
   builder.addParticleChild(0, 1);
   builder.addParticleChild(0, 2);
   builder.addParticleChild(1, 3);
   builder.addParticleChild(2, 3);
 
-  builder.addHit(truth::HitChannel::HGCalCalo, 103, /*detId=*/50, /*energy=*/2.0f, /*recHitIndex=*/0);
+  builder.addHit(truth::HitChannel::HGCalCalo, 0, 103, /*detId=*/50, /*energy=*/2.0f, /*recHitIndex=*/0);
 
   auto index = builder.finish();
 

